### DISCUSSION TITLE
wkhtmltopdf Authentication

### DISCUSSION THREAD

**Discussion by MxRaccoon** [serenity-is/Serenity#7156] [Help]
Hello everyone,  I know that this question is not a problem on the part of Serenity but of Authentication in IIS.

I come from working on a Windows Server 2012 with IIS 8 and a version of Serenity in Framework 4.5, everything worked fine but I had to migrate to Windows 2019 with IIS 10 and my only problem so far is that I receive a timeout when issuing a report with wkhtmltopdf and it is due to some type of Authentication problem since when this configuration is moved in IIS the report is issued on the screen.

I currently have Anonymous Authentication and Forms Authentication enabled on the site. Has anyone gone through something similar that can give me an idea of ​​what configuration I need? The Application Pool user has the necessary permissions as well as the IUSR. 

I appreciate your support. I have already had two full days reading documentation, posts on StackOverflow and doing tests and I can't find the answer.

Thanks in advance

**Comment by volkanceylan** [comment]
Assuming you are using an old version, and all other is same, not clear what the issue is. Are you sure you installed wkhtml e.g. its c++ runtime etc. using its installer? I dont think its auth issue. You may try running wkhtml from console with the same login iis uses.

**Reply by MxRaccoon** [reply]
hi  @volkanceylan , thanks for the comment, yes I am using and old version of serenity, I have the exe file on app_data and bin directories, installed the c++ runtime 2013. Even I ran the project in the server to verify it worked on development enviroment and it does as expected so nothing is missing for it to run. 

If I delete the exe files it throws the correct error, when the files are there it executes and hangs (until timeout), remains open in the task manager and i need to kill the process, so something is blocking the exe to return the file (i guess).

I just uploaded a 6.7 version of serenity in the same server and the tool works fine with that project, so I am starting to think that might be a compatibility issue of some kind, but I am an industrial engineer not a developer an not an expert in IIS or C# (just a serenity enthusiast willing to go to starsharp in the short term) so just taking guesses.

What is the its installer? maybe I am missing something. I just publish the application in my PC and upload it to the server manually.

**Comment by volkanceylan** [comment]
https://wkhtmltopdf.org/

Download install and try running from console for google or your site as shown in the home page.

**Reply by MxRaccoon** [reply]
Hi again, I did install it in my tests but didnt try to run it in console, it did work without issues at first try

![image](https://github.com/serenity-is/Serenity/assets/155673466/4571ede9-ad07-4f09-846b-c4f8fbe0f363)

So i did test on the exe used in the app and again it worked as expected, but when the request comes from the app something is blocking the exe to deliver the file, I can see the wkhtmltopdf pop in the task manager, if i kill the task the report trows the "Generated PDF file is empty!" error if I wait is timeout error😢.

Spent another 3 hours looking, granted permission to apppool user to the CMD, same result.

My net core 6 app runs fine and shows the reports, so I have run out of ideas?

thanks to take your time to read and help me

**Reply by VictorTomaili** [reply]
Can you check the exe for blocking
![image](https://github.com/serenity-is/Serenity/assets/2023440/b240b99d-2ac7-400c-b914-c2e03c2b564f)

Try to set `SiteExternalUrl` in your appsettings.json with the url which can be accessible from your windows server. 
Check the windows event logs for any record related to WKHtmlToPdf
Check the report output directory permissions with the IIS User too, the WKHtmlToPdf uses temp file name `Path.GetTempFileName()` as output of the file. 
You can copy the `WKHtmlToPdfConverter.cs` and `WKHtmlToPdf.cs` files from the source code in serenity repository and update the namespace and name of the classes. Ensure the version should be the same as your application. Register them to services before the `services.AddReporting();` line in your `Startup.cs`
```cs
        services.TryAddSingleton<IWKHtmlToPdfConverter, MyWKHtmlToPdfConverter>();
        services.TryAddSingleton<IHtmlToPdfConverter, MyWKHtmlToPdfConverter>();
```
In `MyWKHtmlToPdf.cs` you can log the `commandLineArgs` parameter to check for any issue in parameters. Try to run the WKHtmlToPdf with the logged parameters and see if the WKHtmlToPdf executable is hanging or not.

**Reply by MxRaccoon** [reply]
hi @VictorTomaili , thanks for the help

The issue is on an old serenity app .net 4.5 framework

The exe file is not blocked, does not have that option in properties

In the wondows event log I only see the errors that are displayed on the browser, like the timeout or the report not found (when the exe process is killed)

You refer as making a console app with those files/codes to try to debug the app?

The path in my local machine is "http://localhost:55558/Report/Render?key=Module.Pedido&opt=%7B%22PedidoID%22%3A8254%7D&print=1" I so it is in the root folder of the published app, I gave permission to the AppPool user and IUSR to the root (and all the files inside)

**Reply by VictorTomaili** [reply]
I wasn't referring to creating a console application. What I meant was that you can take these classes and incorporate them into your project, registering them in the startup.cs file. Afterward, you can alter the code to log particular data, such as console parameters. While localhost may function on local machines, you should verify if your application is reachable via a localhost URL on your server. Additionally, a https or http issue could lead to an endless redirect loop, resulting in WKHtmlToPdf becoming unresponsive. It might be necessary to configure the `SiteExternalUrl` in your appsettings.json.

But your case is different because the .net 4.5 doesn't have appsettings so you can use `<add key="EnvironmentSettings" value="{ SiteExternalUrl : 'http:\\myexternalurl' }" />` in your web.config
To log parameters etc. You can copy the `HtmlToPdfConverter` to your project and log the args to try in console. 
If you think it is related to the folder permission, you can change the line `var tempFile = Path.GetTempFileName();` with your custom path to try permissions. 

https://github.com/serenity-is/Serenity/blob/79903b72d55571aba8452899b694af461c550250/Serenity.Web/Reporting/Tools/HtmlToPdfConverter.cs

**Reply by MxRaccoon** [reply]
Hi, @VictorTomaili , thanks again.

I integrated those files and they work on my developmet app, changed tempfile path to the server app root, same happens (works locally in debug), If I include the Environment settings it returns the empty pdf error, curious thing is that when the render is hanging and I save a change (remove the Environment settings) from the webconfig the report reders so the Idea of an enless direct loop issue may be happening... need to resolve how to write a log from the HtmlPdfConverter args when the app is published to see what is been send to the exe and as you say try to replicate it.

In the mean while I am migrating the most critical process to the net core 6 version thats is working, it looks like whould be more time efficient than trying to debug, but I will sill try to solve it, there are other apps that need pdf functionality and are bigger than the one I am working on, but once migrated all the apps I might go the starsharp way (I have some months thingking about it, have some questions but will email them later) and keep everiything on the latest stable versions.

Will come back with my results on the log. thanks again

**Comment by MxRaccoon** [comment]
Hi! just comming back to give feedback.

I was able to find the error, not serenity related or IIS.... there was an old process installed that was causing a background error, giving some kind of disconection and blocking the pdf rendering.

Thanks for the help, everything helped me to go in the right path to find the issue.

Now is time to continue migrating my apps to the new versions.

Have a great day!
